interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,353 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+)
+
+type componentType int
+
+// Component types in order from lowest to highest precedence.
```

Comment:
Can you please add a paragrah here that explains why the priority is as it is. This can be copy-pasted from the spec.

Reply:
Added one that's rather verbose, please let me know what you think.

Reply:
we need to find a way to structure it such that is readable at the moment it is not. Also it's not clear to me why foo/** has higher precedence over foo/* or maybe it's not true but is not clear reading. It's because of the separator. The way we mix separator in or not in other things it's a bit confusing